Custom extensions cannot import models/widgets/widget

As part of panel we are shipping various custom models, generally that works perfectly but it seems when I export a plot containing custom models to a static HTML file it fails to resolve certain modules, specifically this line raises errors:

          var widget_1 = require("models/widgets/widget"); 

while other imports seem to work just fine, e.g.:

          var p = require("core/properties");

I'm really not quite sure how best to provide a minimal reproducible example, so [here](http://assets.holoviews.org/temp/Panel.html) is the exported HTML file, [here](https://github.com/pyviz/panel/blob/master/panel/models/audio.ts) is the custom model definition and here is the traceback:

```
Uncaught Error: Cannot find module 'models/widgets/widget'
    at s (Panel.html:93)
    at Object.custom/panel.models.widgets.audio (Panel.html:358)
    at s (Panel.html:93)
    at Object.custom/main (Panel.html:147)
    at s (Panel.html:93)
    at Object.t.Bokeh.r.register_plugin (Panel.html:93)
    at outer (Panel.html:137)
    at Panel.html:142
    at Panel.html:132
    at Panel.html:133
```

This may or may not be related to this error I get when I try to import the Widget model like this: 

```
import {Widget} from "models/widgets"
```

```
Compilation failed:

../../../bokeh/bokeh/server/static/js/types/models/widgets/date_picker.d.ts:3:26 - error TS2307: Cannot find module 'pikaday'.

3 import * as Pikaday from "pikaday";
```

>  error TS2307: Cannot find module 'pikaday'.

This is a bug. All those external libraries shouldn't be part of the public API. I though I already fix all instances.

Okay, I removed the DatePicker widget from the ``widgets/index.ts`` which got rid of the pikaday issue but it's still failing to resolve the module in the exported HTML:

```
 Cannot find module 'models/widgets/index'
```

--- a/bokeh/util/compiler.py
+++ b/bokeh/util/compiler.py
@@ -597,7 +597,9 @@ def _bundle_models(custom_models):
                     extra_modules[sig] = True
                     modules.append((sig, code, deps_map))
             else:
-                raise RuntimeError("no such module: %s" % module)
+                index = module + ("" if module.endswith("/") else "/") + "index"
+                if index not in known_modules:
+                    raise RuntimeError("no such module: %s" % module)
 
         return resolved
 

--- a/bokehjs/make/prelude.ts
+++ b/bokehjs/make/prelude.ts
@@ -20,32 +20,54 @@ ${license}
   var define;
   return (function(modules, aliases, entry, parent_require) {
     var cache = {};
-    var main;
 
-    var require = function(name) {
+    var normalize = function(name) {
+      if (typeof name === "number")
+        return name;
+
       if (name === "bokehjs")
-        return main
+        return entry;
 
-      var id = aliases[name] != null ? aliases[name] : name;
+      var alias = aliases[name]
+      if (alias != null)
+        return alias;
 
-      if (!cache[id]) {
-        if (!modules[id]) {
-          if (parent_require)
-            return parent_require(id)
+      var trailing = name.length > 0 && name[name.lenght-1] === "/";
+      var index = aliases[name + (trailing ? "" : "/") + "index"];
+      if (index != null)
+        return index;
 
-          var err = new Error("Cannot find module '" + name + "'");
-          err.code = 'MODULE_NOT_FOUND';
-          throw err;
-        }
-
-        var module = cache[id] = {exports: {}};
-        modules[id].call(module.exports, require, module, module.exports);
-      }
-
-      return cache[id].exports;
+      return name;
     }
 
-    main = require(entry);
+    var require = function(name) {
+      var mod = cache[name];
+      if (!mod) {
+        var id = normalize(name);
+
+        mod = cache[id];
+        if (!mod) {
+          if (!modules[id]) {
+            if (parent_require)
+              return parent_require(id);
+
+            var err = new Error("Cannot find module '" + name + "'");
+            err.code = 'MODULE_NOT_FOUND';
+            throw err;
+          }
+
+          mod = {exports: {}};
+          cache[id] = mod;
+          cache[name] = mod;
+          modules[id].call(mod.exports, require, mod, mod.exports);
+        } else
+          cache[name] = mod;
+      }
+
+      return mod.exports;
+    }
+
+    var main = require(entry);
     main.require = require;
 
     main.register_plugin = function(plugin_modules, plugin_aliases, plugin_entry) {

--- a/sphinx/source/docs/user_guide/examples/extensions_example_wrapping.py
+++ b/sphinx/source/docs/user_guide/examples/extensions_example_wrapping.py
@@ -17,7 +17,7 @@ TS_CODE = """
 
 import {LayoutDOM, LayoutDOMView} from "models/layouts/layout_dom"
 import {ColumnDataSource} from "models/sources/column_data_source"
-import {LayoutItem} from "core/layout/index"
+import {LayoutItem} from "core/layout"
 import * as p from "core/properties"
 
 declare namespace vis {

